src/Interpreters/Cache/QueryCache.cpp

```diff
@@ -52,7 +56,51 @@ struct HasNonDeterministicFunctionsMatcher
     }
 };
 
+struct HasSystemTablesMatcher
+{
+    struct Data
+    {
+        const ContextPtr context;
+        bool has_system_tables = false;
+    };
+
+    static bool needChildVisit(const ASTPtr &, const ASTPtr &) { return true; }
+
+    static void visit(const ASTPtr & node, Data & data)
+    {
+        if (data.has_system_tables)
+            return;
+
+        String database_table; /// or whatever else we get, e.g. just a table
+
+        /// SELECT [...] FROM <table>
+        if (const auto * table_identifier = node->as<ASTTableIdentifier>())
+        {
+            database_table = table_identifier->name();
+        }
+        /// SELECT [...] FROM clusterAllReplicas(<cluster>, <table>)
+        else if (const auto * identifier = node->as<ASTIdentifier>())
+        {
+            database_table = identifier->name();
+        }
+        /// Handle SELECT [...] FROM clusterAllReplicas(<cluster>, '<table>')
+        else if (const auto * literal = node->as<ASTLiteral>())
+        {
+            const auto & value = literal->value; /// (*)
+            database_table = applyVisitor(FieldVisitorDump(), value);
+        }
+
+        /// (*) returns table in quotes, so we can't use .starts_with() for matching
+        static const re2::RE2 is_system_table(String(DatabaseCatalog::TEMPORARY_DATABASE)
+                                    + "|" + DatabaseCatalog::SYSTEM_DATABASE
+                                    + "|" + DatabaseCatalog::INFORMATION_SCHEMA
+                                    + "|" + DatabaseCatalog::INFORMATION_SCHEMA_UPPERCASE);
+        data.has_system_tables = re2::RE2::PartialMatch(database_table, is_system_table);
```

Comment:
seems it will work incorrectly if for example a table name contains a word "system" ? Looks like quite a common case. Should we better use something like `parseDatabaseAndTableName` here to distinguish between table and database? (similar to `Parsers/parseDatabaseAndTableName.cpp` but for string, or may be there is some function like there somewhere else idk)

Reply:
Makes sense, thanks. Done.